source/memory.tex

```diff
@@ -5802,31 +5802,55 @@
 are general-purpose memory resources having the following qualities:
 \begin{itemize}
 \item
-Each resource frees its allocated memory on destruction,
-even if \tcode{deallocate} has not been called for some of the allocated blocks.
+A pool resource has an \defn{upstream memory resource},
+supplied on construction, which is used to allocate memory that
+the pool resource subdivides to satisfy calls to its own
+\tcode{do_allocate} member.
+
+\item
+A \defn{chunk} is a region of storage allocated
+from the upstream memory resource. A pool resource divides
+each chunk into one or more blocks of the same block size.
+Different chunks can have different block sizes, and different
+numbers of blocks.
+
+\item
+A \defn{pool} is a collection of chunks that all have the same
+block size. Each pool resource contains a collection of pools
+with different block sizes.
+
+\item
+On any call to \tcode{release()}, a pool resource frees all the memory
+it obtained from its upstream memory resource by calling
+the upstream resource's \tcode{deallocate} function.
```

Comment:
`it has obtained`
Abbreviating as `the upstream resource's` here feels just a little weird to me, since the true name is `the upstream memory resource's` and if we're aiming for brevity we could just say `the upstream's`.

I admit that my eye caught on this paragraph originally because it refers to "the memory
it [has] obtained from its upstream memory resource by calling the upstream resource's deallocate function"; when in fact that memory was obtained by calling the upstream's _allocate_ function! But I have no suggested fix for that grammar quirk, so my brain found other nits.

Reply:
How about:

> On a call to `release()`, a pool resource calls its upstream memory resource's `deallocate` function to return all the memory that was obtained from its upstream memory resource.

Reply:
Or "that was obtained from upstream".

Reply:
Minor tweak:

> On a call to `release()`, a pool resource calls its upstream memory resource's `deallocate` function to return all the memory that _has been_ obtained from its upstream memory resource.

Or "all the _outstanding_ memory," but that's a little jargony. Could even say "all the _chunks_ that have been obtained," if that's correct. Can't say "_has ever been_ obtained," because we really do mean only the outstanding chunks, not ones that were allocated before the last call to `release()` nor ones that have already been deallocated via `deallocate`.

Wildly ambitious rewrite:

> On a call to `release()`, a pool resource calls its upstream memory resource's `deallocate` function on every chunk that has not yet been deallocated.

Reply:
But it doesn't say all memory that has *ever* been allocated, and we don't bother making that distinction for containers freeing memory with an allocator.